doc/comptime.org

```diff
@@ -0,0 +1,495 @@
+The example below is based on [[https://zig.guide/language-basics/comptime][the zig guide]] [cite:@ComptimeZigGuide2024], it's shows that it can define a new type with a function.
+#+begin_src zig :imports '(std) :main 'no :testsuite yes
+  fn Matrix(
+      comptime T: type,
+      comptime width: comptime_int,
+      comptime height: comptime_int,
+  ) type {
+      return [height][width]T;
+  }
+
+  fn Matrix3D(
+      comptime T: type,
+  ) type {
+      return Matrix(T, 3, 3);
+  }
+
+  test "returning a type" {
+      try std.testing.expect(Matrix(f32, 4, 4) == [4][4]f32);
+  }
+
+  test "returning a 3D matrix" {
+      try std.testing.expect(Matrix3D(f32) == [3][3]f32);
+  }
+
+#+end_src
+
+#+RESULTS:
+
+** Metaprogramming
+*** ~@TypeOf~
+The ~@TypeOf~ builtin function can be used to take as a parameter an expression and return a type.
+
+*** ~@typeInfo~
+This built-in function provides type reflection, it returns information on type.
+
+See the example ~Example with a custom CSV writer based on type~ to have a view of the usability.
+
+** TODO Optimization
+inline for / while
+
+*** How works log in Zig
+In the C language, a common use to use debug print is with Marco.
+Like in this example, if the ~DEBUG~ is defined to ~1~ the code will print the debug info.
+If the ~DEBUG~ is not set, at the compilation, all the print information will be removed.
+#+begin_src C :main 'no :includes <stdio.h>
+
+  #define DEBUG 1
+
+  #if DEBUG 
+  #define TRACE(x) printf x;
+  #else
+  #define TRACE(x)
+  #endif
+
+  int main() {
+    TRACE(("Hello World! : %d\n", 12));
+  }
+#+end_src
+
+#+RESULTS:
+: Hello World! : 12
+
+In Zig, logging uses this same principle, a message level is set at the start of the program (at compile-time) and if the log is not enabled, all the code about the print is removed.
+However, if the log level is greater than the limit, the message will be printed.
+
+The code below shows an extract of the standard library for logging.
+#+begin_src zig
+ fn log(
+    comptime message_level: Level,
+    comptime scope: @Type(.EnumLiteral),
+    comptime format: []const u8,
+    args: anytype,
+) void {
+    if (comptime !logEnabled(message_level, scope)) return;
+
+    std.options.logFn(message_level, scope, format, args);
+}
+#+end_src
+
+In addition, Zig provides some helper functions for logging, such as :
+- ~std.log.debug~
+- ~std.log.info~
+- ~std.log.warn~
+- ~std.log.err~
+
+And if the release mode is set to ~Debug~, the debug log will be printed.
+But if the release mode is set to ~Release*~, the debug log will not print, there is no need to configure the logging to have this behavior.
+
+** Generic data structures
+To create a generic data structure, the same pattern is used as a comptime parameter.
+A function needs to return an anonymous struct as a type ~type~.
+
+In a generic data structure, the ~@This()~ is used to get the type of the data structure because she is anonymous.
```

Comment:
english grammar: im pretty sure a "data structure" has no gender and must therefore be "it" instead of "she" :)